lister.hh

```diff
@@ -153,6 +154,41 @@ private:
     future<directory_entry> guarantee_type(directory_entry de);
 };
 
+class directory_lister {
+    fs::path _dir;
+    lister::dir_entry_types _type;
+    lister::filter_type _filter;
+    lister::show_hidden _do_show_hidden;
+    seastar::queue<std::optional<directory_entry>> _queue;
+    std::unique_ptr<lister> _lister;
+    std::optional<future<>> _done;
+public:
+    directory_lister(fs::path dir,
+            lister::dir_entry_types type = {directory_entry_type::regular, directory_entry_type::directory},
+            lister::filter_type filter = [] (const fs::path& parent_dir, const directory_entry& entry) { return true; },
+            lister::show_hidden do_show_hidden = lister::show_hidden::yes) noexcept
+        : _dir(std::move(dir))
+        , _type(type)
+        , _filter(std::move(filter))
+        , _do_show_hidden(do_show_hidden)
+        , _queue(512 / sizeof(std::optional<directory_entry>))
+    { }
+
+    ~directory_lister();
+
+    // Caller should drain all entries using get()
+    // until it gets a disengaged result or an error.
+    future<std::optional<directory_entry>> get();
+
+    // Abort the directory_lister with a given exception
+    // Further calls to get() will fail with this exception.
+    void abort(std::exception_ptr ex);
+
+    // Close the directory_lister, ignoring any errors.
+    // Must be called if get() did not drain the queue.
```

Comment:
That's correct.

`close()` never returns an error.

The main job of `close()` is to wait for the underlying lister to be `done()`,
anything else is a bonus.

This is needed, as I wrote in the comment, in case 'get()` did not drain the queue, until it got a disengaged `std::optional<directory_entry>`.  If it did then `close()` is not required.
The motivation to have it is when one lists a directory until some specific entry is found, then close() can be used to skip over anything else beyond this entry, without having to wait for all entries to be listed.

The call to `seastar::queue::abort()` will indeed abort any waiting `get()`.